contributors/design-proposals/storage/csi-migration.md

```diff
@@ -0,0 +1,350 @@
+CSI driver in kubelet must be able to find the VolumeAttachment too to get
+AttachmentMetadata for NodeStage/NodePublish.
+
+In downgrade scenario where the migration is then turned off we will have to
+remove these floating VolumeAttachment objects, the same issue is outlined above
+in the Non-Dynamic Provisioned Volumes section.
+
+
+## Interactions with PV-PVC Protection Finalizers
+PV-PVC Protection finalizers prevent deletion of a PV when it is bound to a PVC,
+and prevent deletion of a PVC when it is in use by a pod.
+
+There is no known issue with interaction here. The finalizers will still work in
+the same ways as we are not removing/adding PV’s or PVC’s in out of the ordinary
+ways.
+
+## Dealing with CSI Driver Failures
+Plugin should fail if the CSI Driver is down and migration is turned on. When
+the driver recovers we should be able to resume gracefully.
+
+We will also create a playbook entry for how to turn off the CSI Driver
+migration gracefully, how to tell when the CSI Driver is broken or non-existent,
+and how to redeploy a CSI Driver in a cluster.
+
+## General Upgrade/Downgrade (Migration On/Off) Considerations
+
+What happens when we downgrade to a version without migration after we already
+had migration turned on. Could we have both the internal and external components
+operating on the same PVs or PVCs? Especially for the PV with In-tree plugin
+source and a CSI Source in the annotation. The short answer for non-dynamically
+provisioned volumes is no, since all operations are started directly by the
+attach-detach controller or the kubelet.
+
+This type of race can only occur when CSI and in-tree components are acting in
+reaction to an outside object. This happens at two points: if downgrade is
+performed after a volumeattachment but before the volume is actually attached,
+and if PVC is created for dynamically provisioned and cluster is downgraded
+before provisioning.  In both cases the in-tree plugin and CSI Driver may race
+by both reacting to the external object without knowledge of each other.
+
+For Provisioning this is solved by having the CSI Driver have a different
+provisioner name than the in-tree provisioner. A PVC can only be marked for
+provisioning for one of them at a time.
+
+The VolumeAttachment case is a little more tricky as volumes attached by in-tree
+plugin won’t have VolumeAttachment instance after upgrade. CSI plugin will try
+to attach the volume, create VolumeAttachment and call CSI driver. It’s
+important that the CSI driver is idempotent and the external attacher uses the
+same parameters as the internal volume plugin used to ensure the idempotency.
+Especially CSI VolumeHandle must be carfully chosen! Volumes attached by CSI
+driver will have VolumeAttachment. In-tree volume plugin should be able to
+detach the volume in the usual way after downgrade, however, A/D controller
+should delete the VolumeAttachment.
```

Comment:
Migration in-tree -> csi:

I don't really understand this exact case, Kubelet does not need AttachmentMetadata to mount device, I think it will need `node.status.volumesattached.devicePath` like you mentioned in the second part though, so the reasoning is the same there.

Another problem if we Attach using in-tree, then try to Detach with CSI, we will not have the VolumeAttachment object to use to "signal" the Detach operation. In this case, I think your solution is correct that we must somehow "re-attach" using CSI to create this VolumeAttachment object, and it should be idempotent by definition of the spec (we must ensure we use same device name).

Am I misunderstanding?

CSI -> in-tree:

This one is tricky, I realized I never thought of this case because GCE PD doesn't get device paths from `node.status` for mount operations (it does for UnmountDevice though) but I just looked through AWS and that driver does. 

The problem is that "node.status.volumesAttached.devicePath empty (-> some error)." wont be true, the devicePath will be filled with the csiAttachID. The race is a little more confusing at this point because the in-tree plugin will try to do operations using that, we will have to confirm that it will try to retry later.


So the AI to add to the design:
* A/D Controller should check if each attached device was attached with in-tree or CSI
* If it was attached by the "other" mode, we kick off attach again
* CSI plugin must have idempotent attach (already true), we must make sure in-tree attach is idempotent (TBD)
* Kubelet must retry MountDevice even if `node.status.volumesAttached.devicePath` is filled but incorrect

This is specifically for the case that we switch on/off migration in between Attach and MountDevice/Mount.

We will not worry about what happens when switching between various mount options, or on the detach/Unmount side for now because of the requirement for draining nodes right now. I imagine that's a whole nother can of worms.

Let me know if this makes sense to you.

Reply:
added a section about this

Reply:
Thought about this again. Is this failure mode even possible if we require draining of nodes before flipping the flag on/off on a node? Draining the node would cause all volumes to be detached anyway. Then when Kubelet comes back up with flag on and annotates the node, starting from Attach all the way through the workflow we will be using the same plugin.

Reply:
I guess this could happen if you only decide to flip the flag off on the A/D controller. But that seems strictly in the territory of user error. (mismatched flags on kubelet and A/D Controller)

Reply:
This is actually solved by the combination of adding the NodeAnnotation when the Kubelet is migrated + the requirement to drain nodes before Kubelet Upgrade. We cannot hit these cases.